python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,157 @@
+        super().__init__(self.message)
+
+
+app = FastAPI()
+
+
+@serve.deployment(name="myEventProvider", route_prefix="/event")
+@serve.ingress(app)
+class HTTPEventProvider():
+    """HTTPEventProvider is defined to be a Ray Serve deployment with route_prefix='/event',
+    which will receive external events via an HTTP endpoint. It supports FastAPI, e.g. post.
+    It responds to both poll_for_event() and event_checkpointed() from an HTTPListener instance.
+
+    HTTPListener class is designed to work with the current workflow.wait_for_event()
+    implementation, where an HTTPListener instance will be initiated by the get_message() and
+    message_committed() of the workflow.wait_for_event().
+
+    Example definition
+    ==================
+
+    ```
+    class HTTPEventProvider:
+
+        def __init__(self):
+
+        @app.post("/send_event/{workflow_id}")
+        async def create_event_payload(self, workflow_id: str, req: Request):
+
+        async def get_event_payload(self, workflow_id, event_key):
+
+        async def report_checkpointed(self, event, confirmation):
+
+    ```
+
+    Example Usage
+    =============
+    >>> from ray.workflow.http_event_provider import HTTPEventProvider, HTTPListener
+    >>> ray.init(address='auto', namespace='serve')
+    >>> serve.start(detached=True)
+    >>> HTTPEventProvider.deploy()
+    >>> event_step = workflow.wait_for_event( # doctest: +SKIP
+    ...     HTTPListener, event_key='')
+    >>> handle_event = ... # doctest: +SKIP
+    >>> handle_event.bind(event_step).run() # doctest: +SKIP
+    >>>
+    >>> On a separate python process, it sends an event to the HTTPEventProvider.
+    >>> import requests
+    >>> resp = requests.post('http://127.0.0.1:8000/event/send_event/{workflow_id}',
+    ...     json={'event_key':'my_key','event_payload':'testMessage'})
+    >>>
+
+    """
+
+    def __init__(self):
+        self.event_key_payload: Dict[str, Dict[Any, Any]] = {}
+        self.payload_record: Dict[Any] = {}
+
+    @app.post("/send_event/{workflow_id}")
+    async def create_event_payload(self, workflow_id: str, req: Request):
+        req_json = await req.json()
+        event_key = req_json['event_key']
+        event_payload = req_json['event_payload']
+        if workflow_id not in self.event_key_payload.keys():
+            raise HTTPException(status_code=404, detail="workflow_id not matching")
+        if event_key not in self.event_key_payload[workflow_id].keys():
+            raise HTTPException(status_code=404, detail="event_key not matching")
+        self.event_key_payload[workflow_id][event_key].set_result(event_payload)
+
+        self.payload_record[event_payload] = asyncio.Future()
+        confirmed = await self.payload_record[event_payload]
+        self.payload_record.pop(event_payload)
+        if confirmed:
+            return "ACK_OK: event received and checkpointed successfully."
```

Comment:
Let's follow the google's style for response here: https://google.github.io/styleguide/jsoncstyleguide.xml#data

Reply:
Updated. Please review.